pkg/client/unversioned/portforward/portforward_test.go

```diff
@@ -295,55 +192,121 @@ func TestGetListener(t *testing.T) {
 	}
 }
 
+// fakePortForwarder simulates port forwarding for testing. It implements
+// kubelet.PortForwarder.
+type fakePortForwarder struct {
+	lock sync.Mutex
+	// stores data received from the stream per port
+	received map[uint16]string
+	// data to be sent to the stream per port
+	send map[uint16]string
+}
+
+var _ kubelet.PortForwarder = &fakePortForwarder{}
+
+func (pf *fakePortForwarder) PortForward(name string, uid types.UID, port uint16, stream io.ReadWriteCloser) error {
+	defer stream.Close()
+
+	var wg sync.WaitGroup
+
+	// client -> server
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		// copy from stream into a buffer
+		received := new(bytes.Buffer)
+		io.Copy(received, stream)
+
+		// store the received content
+		pf.lock.Lock()
+		pf.received[port] = received.String()
+		pf.lock.Unlock()
+	}()
+
+	// server -> client
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		// send the hardcoded data to the stream
+		io.Copy(stream, strings.NewReader(pf.send[port]))
+	}()
+
+	wg.Wait()
+
+	return nil
+}
+
+// fakePortForwardServer creates an HTTP server that can handle port forwarding
+// requests.
+func fakePortForwardServer(t *testing.T, testName string, serverSends, expectedFromClient map[uint16]string) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		pf := &fakePortForwarder{
+			received: make(map[uint16]string),
+			send:     serverSends,
+		}
+		kubelet.ServePortForward(w, req, pf, "pod", "uid", 0)
+
+		for port, expected := range expectedFromClient {
+			actual, ok := pf.received[port]
+			if !ok {
+				t.Errorf("%s: server didn't receive any data for port %d", testName, port)
+				continue
+			}
+
+			if expected != actual {
+				t.Errorf("%s: server expected to receive %q, got %q for port %d", testName, expected, actual, port)
+			}
+		}
+
+		for port, actual := range pf.received {
+			if _, ok := expectedFromClient[port]; !ok {
+				t.Errorf("%s: server unexpectedly received %q for port %d", testName, actual, port)
+			}
+		}
+	})
+}
+
 func TestForwardPorts(t *testing.T) {
-	testCases := []struct {
-		Upgrader *fakeUpgrader
-		Ports    []string
-		Send     map[uint16]string
-		Receive  map[uint16]string
-		Err      bool
+	//flag.Lookup("v").Value.Set("9")
```

Comment:
It will be gone before this PR is ready for merging :-)

On Fri, Sep 25, 2015 at 1:06 PM, krousey <notifications@github.com> wrote:

> In pkg/client/unversioned/portforward/portforward_test.go
> <https://github.com/kubernetes/kubernetes/pull/12283#discussion_r40451966>
> :
>
> >  func TestForwardPorts(t *testing.T) {
> > -	testCases := []struct {
> > -		Upgrader *fakeUpgrader
> > -		Ports    []string
> > -		Send     map[uint16]string
> > -		Receive  map[uint16]string
> > -		Err      bool
> > +	//flag.Lookup("v").Value.Set("9")
>
> Remove this line please.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12283/files#r40451966>.
>